src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt

```diff
@@ -0,0 +1,115 @@
+
+  init {
+    intakeSparkMax.restoreFactoryDefaults()
+    intakeSparkMax.clearFaults()
+    armSparkMax.restoreFactoryDefaults()
+    armSparkMax.clearFaults()
+
+    // TODO(check if this is right)
+    intakeSparkMax.enableVoltageCompensation(
+      ManipulatorConstants.INTAKE_VOLTAGE_COMPENSATION.inVolts
+    )
+    intakeSparkMax.setSmartCurrentLimit(ManipulatorConstants.INTAKE_SUPPLY_CURRENT_LIMIT)
+    intakeSparkMax.setIdleMode(CANSparkMax.IdleMode.kCoast)
+    intakeSparkMax.inverted = ManipulatorConstants.INTAKE_MOTOR_INVERTED
+    intakeSparkMax.burnFlash()
+    intakeSparkMax.openLoopRampRate = ManipulatorConstants.INTAKE_RAMP_RATE
+
+    armSparkMax.enableVoltageCompensation(ManipulatorConstants.ARM_VOLTAGE_COMPENSATION.inVolts)
+    armSparkMax.setSmartCurrentLimit(ManipulatorConstants.ARM_SUPPLY_CURRENT_LIMIT)
+    armSparkMax.inverted = ManipulatorConstants.ARM_MOTOR_INVERTED
+    armSparkMax.burnFlash()
+    armSparkMax.openLoopRampRate = ManipulatorConstants.ARM_RAMP_RATE
+  }
+
+  override fun setRollerPower(percentOutput: Double) {
+    intakeSparkMax.set(percentOutput)
+  }
+
+  override fun updateInputs(inputs: ManipulatorIO.ManipulatorIOInputs) {
+    inputs.rollerPosition = intakeSensor.position
+    inputs.rollerVelocity = intakeSensor.velocity
+    inputs.rollerAppliedOutput = intakeSparkMax.appliedOutput
+    inputs.rollerStatorCurrent = intakeSparkMax.outputCurrent.amps
+    // BatteryVoltage * SupplyCurrent = AppliedVoltage * StatorCurrent
+    // AppliedVoltage = percentOutput * BatteryVoltage
+    // SuplyCurrent = (percentOutput * BatteryVoltage / BatteryVoltage) * StatorCurrent =
+    // percentOutput * statorCurrent
+    inputs.rollerSupplyCurrent = inputs.rollerStatorCurrent * intakeSparkMax.appliedOutput
+    inputs.rollerTempCelcius = intakeSparkMax.motorTemperature.celsius
+
+    inputs.armPosition = armSensor.position
+    inputs.armVelocity = armSensor.velocity
+    inputs.armAppliedOutput = armSparkMax.appliedOutput
+    inputs.armStatorCurrent = armSparkMax.outputCurrent.amps
+    // BatteryVoltage * SupplyCurrent = AppliedVoltage * StatorCurrent
+    // AppliedVoltage = percentOutput * BatteryVoltage
+    // SuplyCurrent = (percentOutput * BatteryVoltage / BatteryVoltage) * StatorCurrent =
+    // percentOutput * statorCurrent
+    inputs.armSupplyCurrent = inputs.armStatorCurrent * armSparkMax.appliedOutput
+    inputs.armTempCelcius = armSparkMax.motorTemperature.celsius
+  }
+  override fun setOpenLoop(percentOutput: Double) {
+    armSparkMax.set(percentOutput)
+  }
+
+  override fun setPosition(position: Length, feedforward: ElectricalPotential) {
+    armPIDController.setFF(feedforward.inVolts)
+  }
+
+  override fun zeroEncoder() {
+    armSparkMax.encoder.position = 0.0
+  }
+
+  override fun configPID(
+    kP: ProportionalGain<Meter, Volt>,
+    kI: IntegralGain<Meter, Volt>,
+    kD: DerivativeGain<Meter, Volt>
+  ) {
+    armPIDController.p = kP.inVoltsPerMeter
```

Comment:
This is not correct -- the Spark MAX PID controller does not take PID values in these units. You need to use your linear mechanism sensor to convert the PID units to native units

Reply:
an example can be found in drivetrain code in SwerveIOFalcon init